### ISSUE TITLE
Platforming collision bug in combination with "player is jumping"

### ISSUE THREAD

**Issue by Wackytoaster** [Scirra/Construct-3-bugs#1453]
## Problem description

I noticed some odd behavior though I´m not certain where it actually comes from. Basically I have a player sprite with platformer behavior and a sprite to detect collisions for the head (think ala Mario hitting a block) The sprite is pinned to the head and sticks a little over the player sprite.

Now I check if the head collider is colliding with the block AND if the player is currently jumping. But nothing happens.  When I step through in the debugger, it definitely shows that a collision should be detected AND that the player is currently jumping, the event does nothing though.

Stuff I tried

- When I disable the check for "player is jumping" it does work fine.
- It also works if I give the head collider sprite a rather big height. (but I don´t want to "hit" the block before the player actually does hit it)
- Exchanging the pin behavior with setting the head collider to the player every tick, still doesn´t work.
- The same results happen without a head collider, and instead check if the player is colliding with the block. At first I thought that probably the player changes from jumping to something else on impact, but stepping through in the debugger shows it only switches to falling after the collision. Also disabling the check for "jumping" solves the issue.
- Using "overlapping" instead of collision also does not work.
EDIT: Comparing the players vectorY instead of "is jumping" also does not help

EDIT2: I sort of solved this by checking "is player overlapping block at offset" but I still had to disable the "is jumping". While the result now works as intended (i can only hit the block from below), it´s imo not optimal to forego the check if the player is jumping/moving upwards.

## Attach a .c3p

[platformbug.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/1865872/platformbug.c3p.zip)

## Steps to reproduce

0.5 Honestly just check the c3p.
1. Make player with platform behavior (optional: add said head collider sprite)
2. Make a block
3. Check if the player (or collider) is colliding with the block AND if the player is currently jumping

## Observed result

The event will not trigger, though both conditions should be true (as revealed when stepping through with the debugger)

## Expected result

The collision to be detected as expected.

## Affected browsers

* Chrome: yes

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 65.0.3325.181
Browser engine: Blink
Browser architecture: 64-bit
Context: webapp
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1.75
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36
C3 release: r94.2 (stable)
Language setting: en-US

Local storage
Storage quota (approx): 30 gb
Storage usage (approx): 209 mb (0.7%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: no
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GTX 960M Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by Lobo93** [comment]
I don't think this is a bug
I played with your c3p in debug mode a bit and, at the exact frame the collision sprite overlaps the block, the player also hits it, and since the player is hitting the block, its vertical speed is <=0, which means he isn't jumping anymore, so it won't trigger events requiring the "is jumping" condition :(

**Comment by AshleyScirra** [comment]
The question is basically "at the moment the player hits an overhead obstacle, are they jumping or falling?" Firstly, I don't think this question makes sense. It's the exact moment it changes.

If you log the VectorY expression at the moment of collision, the Y vector is 0. So the answer is "neither: they are stopped". I think this makes sense, so closing as not a bug.

**Comment by Wackytoaster** [comment]
> I played with your c3p in debug mode a bit and, at the exact frame the collision sprite overlaps the block, the player also hits it
No it doesn´t? That´s what I also expected first and I ruled it out. Unless there is some quirk with the debugger there is defenitely 1 frame where both conditions are true. There might even be a second frame but that depends on how "is jumping" is actually defined. On collision, the vectorY is 0, but the animation mode is still "jumping", that´s why I´d assume that the player still counts as "jumping". Even if not, 1 frame should still be enough.
[1frame.mp4.zip](https://github.com/Scirra/Construct-3-bugs/files/1883106/1frame.mp4.zip)

Also... _X-files theme starts playing_
[aliens.mp4.zip](https://github.com/Scirra/Construct-3-bugs/files/1883109/aliens.mp4.zip)
Ok, this might just be a debugger quirk.

Also also... when I throttle my CPU with the dev console, it actually starts to work although rather inconsistent. I got the idea when there was a brief lag and it randomly worked.

Not sure what to make out of this, seems to be some kind of edge-case where there either is a frame or there isnt. Anyway, I can work around that, but I still think there is something odd going on. :V
